### ISSUE TITLE
not starting

### ISSUE THREAD

**Issue by user770** [privacore/open-source-search-engine#38]
root@server:/home/open-source-search-engine# ./gb start
admin: ssh 127.0.0.1 '/home/open-source-search-engine/gbstart.sh 0' &
root@server:/home/open-source-search-engine# 

Privacore is not starting (gb start), it's creat sometimes a fatal_error file. Just ./gb is working for around 1 minute then it's crashed. I'm running it on ubuntu, gbstart.sh is executable, it's possible to ssh without password. pdftohtml is installed. Where can I look for errors?

root@server:/home/open-source-search-engine# ./gb
...
20171028-165234-817 0000 016711 INF build: https://perso.example.com/ redirected to  https://www.example.com/
20171028-165234-821 0000 016733 INF db: merge: Map generation succeeded.
20171028-165234-821 0000 016711 INI db: Resuming a killed merge.
20171028-165234-821 0000 016733 ERR Spider.cpp:dedupSpiderdbList:2752: lastKey is null. Should not happen?
20171028-165234-821 0000 016711 INF db: Merge status: Success.
20171028-165234-821 0000 016732 INF db: Writing map spiderdb0000.map.
20171028-165234-821 0000 016732 INF db: Saving spiderdb0000.map
20171028-165234-821 0000 016711 INF merge: Merge succeeded. spiderdb0000.002.0003.dat (#0) has 908458 positive and 0 negative recs.
20171028-165234-821 0000 016711 INF merge: Files had 704207 positive and 0 negative recs.
20171028-165234-821 0000 016711 INF merge: spiderdb gained 204251 positives.
20171028-165234-821 0000 016711 INF merge: Unlinking merged file /home/open-source-search-engine/coll.main.0/spiderdb0001.dat (#1).
20171028-165234-821 0000 016711 INF merge: Unlinking map file spiderdb0001.map (#1).
20171028-165234-822 0000 016731 TRC BigFile.cpp:unlinkWrapper:1832: disk: unlink [/home/open-source-search-engine/coll.main.0/spiderdb0001.dat]
20171028-165234-822 0000 016730 TRC BigFile.cpp:unlinkWrapper:1832: disk: unlink [/home/open-source-search-engine/coll.main.0/spiderdb0001.map]
20171028-165234-827 0000 016711 INF build: https://www.example.com/robots.txt redirected to https://www.example.com/robots.txt
20171028-165234-835 0000 016711 INF merge: Done unlinking all files.
20171028-165234-835 0000 016711 INF db: Renaming spiderdb0000.002.0003.dat of size 135411381 to to final filename(s)
20171028-165234-835 0000 016711 TRC BigFile.cpp:rename:1560: rename: submitting rename of spiderdb0000.map to spiderdb0001.map part #0
20171028-165234-835 0000 016711 TRC BigFile.cpp:rename:1560: rename: submitting rename of spiderdb0000.002.0003.dat to spiderdb0001.dat part #0
20171028-165234-835 0000 016731 TRC BigFile.cpp:renameP1Wrapper:1608: disk: rename [/tmp/gb_merge_space/0/coll.main.0/spiderdb0000.map] to [/home/open-source-search-engine/coll.main.0/spiderdb0001.map]
20171028-165234-835 0000 016730 TRC BigFile.cpp:renameP1Wrapper:1608: disk: rename [/tmp/gb_merge_space/0/coll.main.0/spiderdb0000.002.0003.dat] to [/home/open-source-search-engine/coll.main.0/spiderdb0001.dat]
20171028-165234-836 0000 016730 ERR moveFile:open(/tmp/gb_merge_space/0/coll.main.0/spiderdb0000.002.0003.dat) failed with errno=22 (Invalid argument)
20171028-165234-836 0000 016711 ERR BigFile.cpp:doneP1RenameWrapper:1640: doneRenameWrapper. rename failed: [spiderdb0000.002.0003.dat] [Invalid argument]
20171028-165234-836 0000 016711 ERR Dumping BigFile at 0x7ff01c05d430
20171028-165234-836 0000 016711 ERR m_flags................: 66
20171028-165234-836 0000 016711 ERR m_maxParts.............: 1
20171028-165234-836 0000 016711 ERR m_numParts.............: 1
20171028-165234-836 0000 016711 ERR m_vfd..................: 48
20171028-165234-836 0000 016711 ERR m_fileSize.............: 135411381
20171028-165234-836 0000 016711 ERR m_lastModified.........: 19700101-005959
20171028-165234-836 0000 016711 ERR m_outstandingUnlinkJobCount: 0
20171028-165234-836 0000 016711 ERR m_outstandingRenameP1JobCount: 1
20171028-165234-836 0000 016711 ERR m_outstandingRenameP2JobCount: 0
20171028-165234-836 0000 016711 ERR m_isClosing............: [false]
20171028-165234-836 0000 016711 ERR m_dir..................: (hex dump)
0000:   2F 74 6D 70 2F 67 62 5F   6D 65 72 67 65 5F 73 70    /tmp/gb_merge_sp
0010:   61 63 65 2F 30 2F 63 6F   6C 6C 2E 6D 61 69 6E 2E    ace/0/coll.main.
0020:   30                                                   0               

20171028-165234-836 0000 016711 ERR m_baseFilename.........: (hex dump)
0000:   73 70 69 64 65 72 64 62   30 30 30 30 2E 30 30 32    spiderdb0000.002
0010:   2E 30 30 30 33 2E 64 61   74                         .0003.dat       

20171028-165234-836 0000 016711 ERR m_newBaseFilename......: (hex dump)
0000:   73 70 69 64 65 72 64 62   30 30 30 31 2E 64 61 74    spiderdb0001.dat

20171028-165234-836 0000 016711 ERR m_newBaseFilenameDir...: (hex dump)
0000:   2F 68 6F 6D 65 2F 6F 70   65 6E 2D 73 6F 75 72 63    /home/open-sourc
0010:   65 2D 73 65 61 72 63 68   2D 65 6E 67 69 6E 65 2F    e-search-engine/
0020:   63 6F 6C 6C 2E 6D 61 69   6E 2E 30                   coll.main.0     

20171028-165234-836 0000 016711 ERR g_unlinkRenameThreads..: 1
20171028-165234-836 0000 016711 WRN File.cpp:rollbackMovePhase1:231: disk: trying to rollback rename-phase1 [/tmp/gb_merge_space/0/coll.main.0/spiderdb0000.002.0003.dat] to [spiderdb0000.002.0003.dat/spiderdb0001.dat], unlink() failed with errno=2 (No such file or directory).
20171028-165234-836 0000 016711 DBG rdb: thread completed rename operation for collnum=0 #outstanding_jobs=2
20171028-165234-836 0000 016711 ERR merge: renaming file(s) failed, g_errno=22 (Invalid argument)
20171028-165234-836 0000 016711 ERR gb: Printing stack trace
20171028-165234-836 0000 016731 ERR moveFile:open(/tmp/gb_merge_space/0/coll.main.0/spiderdb0000.map) failed with errno=22 (Invalid argument)
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x5412fb
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x4dc4e7
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x50891a
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x56611d
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x42650f
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x426278
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x461b5c
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x46283d
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x5401da
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x54165d
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x541c25
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x4134e8
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x40f410
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x7ff045d206ba
20171028-165234-836 0000 016711 ERR addr2line -e ./gb 0x7ff04313b3dd
20171028-165234-890 0000 016711 INF gb: Shutting down urgently. Timed try #0.
20171028-165234-890 0000 016711 INF Finalizing wanted-checking
20171028-165234-890 0000 016711 INF Finalized wanted-checking
20171028-165234-891 0000 016711 DBG gb: disabling threads
20171028-165234-893 0000 016711 DBG gb: trying to shutdown
20171028-165234-893 0000 016711 DBG gb: calling save tree for titledb
20171028-165234-893 0000 016711 INF db: Saving /home/open-source-search-engine/titledb-saved.dat
20171028-165235-003 0000 016711 ERR db: Had error saving tree to disk for titledb: Invalid argument.
20171028-165235-003 0000 016711 DBG gb: calling save tree for posdb
20171028-165235-003 0000 016711 INF db: Saving /home/open-source-search-engine/posdb-buckets-saved.dat
20171028-165236-037 0000 016711 ERR db: Had error saving tree to disk for posdb: Invalid argument.
20171028-165236-037 0000 016711 INF db: Saving posdb-saved.idx
20171028-165236-038 0000 016711 INF db: Saved 4813 index keys for posdb
20171028-165236-038 0000 016711 DBG gb: calling save tree for spiderdb
20171028-165236-038 0000 016711 INF db: Saving /home/open-source-search-engine/spiderdb-saved.dat
20171028-165236-082 0000 016711 INF db: Done saving spiderdb with 26980 keys (4322940 bytes)
20171028-165236-082 0000 016711 DBG gb: calling save tree for clusterdb
20171028-165236-082 0000 016711 INF db: Saving /home/open-source-search-engine/clusterdb-saved.dat
20171028-165236-083 0000 016711 INF db: Done saving clusterdb with 18876 keys (509678 bytes)
20171028-165236-083 0000 016711 DBG gb: calling save tree for tagdb
20171028-165236-083 0000 016711 INF db: Saving /home/open-source-search-engine/tagdb-saved.dat
20171028-165236-442 0000 016711 INF db: Done saving tagdb with 321095 keys (24119487 bytes)
20171028-165236-442 0000 016711 DBG gb: calling save tree for linkdb
20171028-165236-442 0000 016711 INF db: Saving /home/open-source-search-engine/linkdb-saved.dat
20171028-165236-472 0000 016711 INF db: Done saving linkdb with 800875 keys (34437651 bytes)
20171028-165236-472 0000 016711 DBG gb: calling save tree for doledb
20171028-165236-472 0000 016711 INF db: Saving /home/open-source-search-engine/doledb-saved.dat
20171028-165236-561 0000 016711 INF db: Done saving doledb with 75111 keys (12049049 bytes)
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 DBG gb: calling save tree for
20171028-165236-561 0000 016711 INF db: Saving /home/open-source-search-engine/coll.main.0waitingtree-saved.dat
20171028-165236-561 0000 016711 INF db: Done saving waitingtree with 1354 keys (611063 bytes)
20171028-165236-562 0000 016711 INF db: Saving rdb indexes
20171028-165236-562 0000 016711 INF db: Saving rdb maps
20171028-165236-562 0000 016711 INF gb: Shutting down udp server port 5998.
20171028-165236-562 0000 016711 INF gb: Closing udp server socket port 5998.
20171028-165236-562 0000 016711 INF gb: Shut down udp server port 5998 successfully.
20171028-165236-562 0000 016711 INF gb: Shutting down udp server port 9000.
20171028-165236-562 0000 016711 INF gb: Closing udp server socket port 9000.
20171028-165236-562 0000 016711 INF gb: Shut down udp server port 9000 successfully.
20171028-165236-562 0000 016711 INI admin: stopping real time profiler
20171028-165236-562 0000 016711 INF dns: Finalizing library
20171028-165236-563 0000 016711 INF gb: Saving miscellaneous data files.
20171028-165236-563 0000 016711 INF db: Saving /home/open-source-search-engine/gb.conf
20171028-165236-563 0000 016711 INF coll: Saving /home/open-source-search-engine/coll.main.0/coll.conf
20171028-165236-564 0000 016711 INF build: Saving /home/open-source-search-engine/addsinprogress.saving
20171028-165236-564 0000 016711 INI db: Saving 3000000 bytes of cache to /home/open-source-search-engine//robots.txt.cache
20171028-165236-565 0000 016711 INI db: Saving 128000 bytes of cache to /home/open-source-search-engine//dns.cache
20171028-165236-566 0000 016711 WRN gb: Dumping core after saving.
Aborted
root@server:/home/open-source-search-engine#

**Comment by arpadlukacs** [comment]
I also have a similar problem on a freshly installed server (Ubuntu 16.04.3) and the search engine was complied from the latest source. It seemed to work after the initial start, then it stopped working and now I could not restart it and I have a similar log which ends with a dumping core line.

**Comment by br-privacore** [comment]
Try to configure the merge drive and don't use /tmp/ - some systems auto-delete files there.

Do a debug build, and attach a stack trace from gdb and we'll have a look.

**Comment by arpadlukacs** [comment]
Ok, thank you. I will do if the issue occurs again. For now, I just reinstalled it looks ok so far, except an error in the log. I will open a new issue for that.

**Comment by arpadlukacs** [comment]
Maybe it worths mentioning here that the search engine will  not start if you have a fatal_error file in it's folder... I found that in a comment in another issue. https://github.com/privacore/open-source-search-engine/issues/19

That does not fix the initial problem that caused to stop with a fatal_error though.

**Comment by br-privacore** [comment]
I am 99.9% sure the original error was caused by having the merge drive configured on the /tmp partition. 

If a fatal error occurs, like above, we shut down and make sure the instance cannot start again until you have manually removed the fatal_error file. This is because the situation is really fatal - you need to manually inspect what went wrong. We prefer that instead of trying to do auto-recovery with potential data loss.

Try it again with the merge drive configured as a non-tmp folder, and re-open the issue if you see the error again.

**Comment by arpadlukacs** [comment]
I switched to another folder for the merge space already and I have not seen the same fatal error since, but I do see other issues.
